File: clients/java/src/test/java/io/volant/JoinGroupMemberTest.java
package io.volant;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.Test;

/** v0.131: thin Client JoinGroup rejoin (records decoded member_id). */
class JoinGroupMemberTest {
    @Test
    void joinGroupSendsEmptyMemberId() throws Exception {
        try (JoinGroupStub srv = new JoinGroupStub()) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                JoinGroupResult j = c.joinGroup("g", List.of("t"), 10_000);
                assertEquals("m-1", j.memberId);
                assertEquals(1L, j.generation);
            }
            srv.assertOk();
            assertEquals("", srv.memberId.get());
            assertEquals("g", srv.group.get());
            assertEquals(Collections.singletonList(Codec.OP_JOIN_GROUP), srv.opcodes);
        }
    }

    @Test
    void joinGroupMemberEncodesId() throws Exception {
        try (JoinGroupStub srv = new JoinGroupStub()) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                JoinGroupResult j = c.joinGroupMember("g", "rejoin-1", List.of("t"), 10_000);
                assertEquals("m-1", j.memberId);
            }
            srv.assertOk();
            assertEquals("rejoin-1", srv.memberId.get());
        }
    }

    @Test
    void joinGroupMemberEmptyMatchesPublicApi() throws Exception {
        try (JoinGroupStub srv = new JoinGroupStub()) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                c.joinGroupMember("g", "", List.of("t"), 10_000);
            }
            srv.assertOk();
            assertEquals("", srv.memberId.get());
        }
    }

    private static final class JoinGroupStub implements AutoCloseable {
        final int port;
        final List<Integer> opcodes = new CopyOnWriteArrayList<>();
        final AtomicReference<String> memberId = new AtomicReference<>();
        final AtomicReference<String> group = new AtomicReference<>();
        private final ServerSocket listen;
        private final Thread thread;
        private final AtomicReference<Exception> error = new AtomicReference<>();
        private final CountDownLatch done = new CountDownLatch(1);

        JoinGroupStub() throws IOException {
            ServerSocket ss = new ServerSocket();
            ss.setReuseAddress(true);
            ss.bind(new InetSocketAddress("127.0.0.1", 0));
            ss.setSoTimeout(5_000);
            this.listen = ss;
            this.port = ss.getLocalPort();
            this.thread = new Thread(
                    () -> {
                        try (Socket conn = listen.accept()) {
                            handle(conn);
                        } catch (Exception e) {
                            error.set(e);
                        } finally {
                            done.countDown();
                        }
                    },
                    "volant-join-group-member");
            this.thread.setDaemon(true);
            this.thread.start();
        }

        void assertOk() throws Exception {
            if (!done.await(5, TimeUnit.SECONDS)) {
                throw new IOException("server did not finish");
            }
            Exception e = error.get();
            if (e != null) {
                throw e;
            }
        }

        @Override
        public void close() {
            try {
                listen.close();
            } catch (IOException ignored) {
                // best-effort
            }
            try {
                thread.join(2_000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }

        private void handle(Socket conn) throws Exception {
            conn.setSoTimeout(5_000);
            InputStream in = conn.getInputStream();
            OutputStream out = conn.getOutputStream();
            byte[] buf = new byte[0];
            while (true) {
                Frame.Decode d = Frame.tryDecode(buf);
                if (d.frame == null) {
                    byte[] tmp = new byte[4096];
                    int n = in.read(tmp);
                    if (n < 0) {
                        return;
                    }
                    byte[] next = new byte[buf.length + n];
                    System.arraycopy(buf, 0, next, 0, buf.length);
                    System.arraycopy(tmp, 0, next, buf.length, n);
                    buf = next;
                    continue;
                }
                buf = d.rest;
                opcodes.add(d.frame.opcode);
                if (d.frame.opcode != Codec.OP_JOIN_GROUP) {
                    throw new IOException("unexpected opcode " + d.frame.opcode);
                }
                Codec.JoinGroupRequest req = Codec.decodeJoinGroupRequest(d.frame.payload);
                group.set(req.groupId);
                memberId.set(req.memberId);
                byte[] payload = Codec.encodeJoinGroupResponse(
                        new Codec.JoinGroupResponse(
                                0, 1L, "m-1", Collections.emptyList(), Collections.emptyList()));
                out.write(Frame.encode(Codec.OP_JOIN_GROUP, d.frame.correlationId, payload));
                out.flush();
            }
        }
    }
}

File: clients/go/join_group_member_test.go
package volant_test

import (
	"net"
	"testing"
	"time"

	volant "github.com/volant-mq/volant/clients/go"
	"github.com/volant-mq/volant/clients/go/codec"
	"github.com/volant-mq/volant/clients/go/frame"
)

// v0.131: thin Client JoinGroup rejoin (records decoded member_id).

type joinGroupMemberResult struct {
	opcodes  []uint16
	memberID string
	group    string
	err      error
}

func serveJoinGroupMember(t *testing.T) (addr string, got *joinGroupMemberResult, stop func()) {
	t.Helper()
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	res := &joinGroupMemberResult{}
	done := make(chan struct{})
	go func() {
		defer close(done)
		conn, err := ln.Accept()
		if err != nil {
			res.err = err
			return
		}
		defer conn.Close()
		_ = conn.SetDeadline(time.Now().Add(5 * time.Second))
		buf := make([]byte, 0, 4096)
		tmp := make([]byte, 4096)
		for {
			f, rest, err := frame.TryDecode(buf)
			if err != nil {
				res.err = err
				return
			}
			if f == nil {
				n, err := conn.Read(tmp)
				if n > 0 {
					buf = append(buf, tmp[:n]...)
				}
				if err != nil {
					res.err = err
					return
				}
				continue
			}
			buf = append([]byte(nil), rest...)
			res.opcodes = append(res.opcodes, f.Opcode)
			if f.Opcode != codec.OpJoinGroup {
				res.err = &frame.ProtocolError{Msg: "unexpected opcode"}
				return
			}
			req, err := codec.DecodeJoinGroupRequest(f.Payload)
			if err != nil {
				res.err = err
				return
			}
			res.group = req.GroupID
			res.memberID = req.MemberID
			payload, err := codec.EncodeJoinGroupResponse(codec.JoinGroupResponse{
				ErrorCode:  0,
				Generation: 1,
				MemberID:   "m-1",
			})
			if err != nil {
				res.err = err
				return
			}
			raw, err := frame.Encode(codec.OpJoinGroup, f.CorrelationID, payload)
			if err != nil {
				res.err = err
				return
			}
			if _, err := conn.Write(raw); err != nil {
				res.err = err
				return
			}
		}
	}()
	return ln.Addr().String(), res, func() {
		_ = ln.Close()
		select {
		case <-done:
		case <-time.After(2 * time.Second):
		}
	}
}

func TestJoinGroupSendsEmptyMemberID(t *testing.T) {
	addr, got, stop := serveJoinGroupMember(t)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	j, err := c.JoinGroup("g", []string{"t"}, 10_000)
	if err != nil {
		t.Fatal(err)
	}
	if j.MemberID != "m-1" || j.Generation != 1 {
		t.Fatalf("result=%+v", j)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.memberID == "" {
		t.Fatal("member_id on the wire is empty")
	}
	if got.group != "g" {
		t.Fatalf("group %q", got.group)
	}
	if len(got.opcodes) != 1 || got.opcodes[0] != codec.OpJoinGroup {
		t.Fatalf("opcodes %+v", got.opcodes)
	}
}

func TestJoinGroupMemberEncodesID(t *testing.T) {
	addr, got, stop := serveJoinGroupMember(t)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	j, err := c.JoinGroupMember("g", "rejoin-1", []string{"t"}, 10_000)
	if err != nil {
		t.Fatal(err)
	}
	if j.MemberID != "m-1" {
		t.Fatalf("member %q", j.MemberID)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.memberID != "rejoin-1" {
		t.Fatalf("member %q want rejoin-1", got.memberID)
	}
}

func TestJoinGroupMemberEmptyMatchesPublicAPI(t *testing.T) {
	addr, got, stop := serveJoinGroupMember(t)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	if _, err := c.JoinGroupMember("g", "", []string{"t"}, 10_000); err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.memberID == "" {
		t.Fatal("member_id on the wire is empty")
	}
}

func TestJoinGroupMemberInstanceEncodesMemberID(t *testing.T) {
	addr, got, stop := serveJoinGroupMember(t)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	if _, err := c.JoinGroupMemberInstance("g", "rejoin-1", []string{"t"}, 10_000, "pod-1"); err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.memberID != "rejoin-1" {
		t.Fatalf("member %q want rejoin-1", got.memberID)
	}
}
